File: decl-py/tests/internal/package_test.py
"""package (tests/internal/checks.json): the manifest reader and the
package hash the lock file rests on."""

from __future__ import annotations

import shutil
from pathlib import Path

from decl.package import package_hash, parse_manifest

ROOT = Path(__file__).resolve().parents[3]


def test_manifest(tmp_path: Path) -> None:
    def read(text: str) -> tuple:
        (tmp_path / "decl.toml").write_text(text, encoding="utf-8")
        codes: list[str] = []
        m = parse_manifest(str(tmp_path / "decl.toml"), lambda c, _m: codes.append(c))
        return m, codes

    good, codes = read('name = "app"\nversion = "1.0.0"\n\n[dependencies]\ncorelib = "1.0.0"\n')
    assert codes == [] and good is not None
    assert (good["name"], good["version"]) == ("app", "1.0.0")
    assert dict(good["dependencies"]) == {"corelib": "1.0.0"}
    _, codes = read(
        'name = "app"\nversion = "1.0.0"\nflavor = "x"\n\n[dependencies]\ncorelib = "^1.0"\n'
    )
    assert "E3011" in codes and "E3012" in codes


def test_hash(tmp_path: Path) -> None:
    corelib = ROOT / "tests/packages/app/decl_modules/corelib"
    locked = (ROOT / "tests/packages/lock/decl.lock").read_text(encoding="utf-8").split()[2]
    h1 = package_hash(str(corelib))
    assert h1 == locked
    assert package_hash(str(corelib)) == h1, "the same on a second call"
    copy = tmp_path / "corelib"
    shutil.copytree(corelib, copy)
    with open(copy / "types/base.decl", "a", encoding="utf-8") as f:
        f.write("// drift\n")
    assert package_hash(str(copy)) != h1, "different for a copy with one file appended to"

File: decl-rs/tests/internal/package_test.rs
//! package: the manifest reader and the package hash the lock file rests on.
use crate::common::{copy_dir, root, temp_dir};
use decl_lang::package::{package_hash, parse_manifest};

#[test]
fn manifest() {
    let dir = temp_dir("manifest");
    let read = |text: &str| {
        std::fs::write(dir.join("decl.toml"), text).unwrap();
        let mut codes: Vec<String> = vec![];
        let m = parse_manifest(&dir.join("decl.toml"), &mut |c, _m| {
            codes.push(c.to_string())
        });
        (m, codes)
    };
    let (good, codes) =
        read("name = \"app\"\nversion = \"1.0.0\"\n\n[dependencies]\ncorelib = \"1.0.0\"\n");
    assert!(codes.is_empty(), "{codes:?}");
    let good = good.expect("a manifest");
    assert_eq!(
        (good.name.as_str(), good.version.as_str()),
        ("app", "1.0.0")
    );
    assert_eq!(
        good.dependencies,
        vec![("corelib".to_string(), "1.0.0".to_string())]
    );
    let (_, codes) = read("name = \"app\"\nversion = \"1.0.0\"\nflavor = \"x\"\n\n[dependencies]\ncorelib = \"^1.0\"\n");
    assert!(
        codes.contains(&"E3011".to_string()) && codes.contains(&"E3012".to_string()),
        "{codes:?}"
    );
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn hash() {
    let corelib = root().join("tests/packages/app/decl_modules/corelib");
    let lock = std::fs::read_to_string(root().join("tests/packages/lock/decl.lock")).unwrap();
    let locked = lock.trim().split(' ').nth(2).unwrap().to_string();
    let h1 = package_hash(&corelib);
    assert_eq!(h1, locked);
    assert_eq!(package_hash(&corelib), h1, "the same on a second call");
    let copy = temp_dir("hash");
    copy_dir(&corelib, &copy);
    let base = copy.join("types/base.decl");
    let mut s = std::fs::read_to_string(&base).unwrap();
    s.push_str("// drift\n");
    std::fs::write(&base, s).unwrap();
    assert_ne!(
        package_hash(&copy),
        h1,
        "different for a copy with one file appended to"
    );
    let _ = std::fs::remove_dir_all(&copy);
}
